### PR TITLE
Block Editor: Fix undo level inconsistencies.

### PR DESCRIPTION
**Pull request by epiqueras**
## Description

This PR fixes a couple of undo level related inconsistencies that surfaced after merging #16932.

In the Block Editor, `withPersistentBlockChange` marks any change to an attribute different than the last, as persistent. However, it considers all actions for this comparison, so an explicit persistence change triggered by a timeout like the one that fires after not typing for a while, would break this "chain of comparison", and the next change would be marked as persistent even if targeted at the same attribute as the one before the explicit persistence change. This resulted in scenarios where typing "Hello", waiting a bit, then typing "World", would result in 4 undo levels instead of 3. Namely, there would be an extra unnecessary undo level in the middle, requiring an extra click to undo. This PR fixes this by ignoring explicit persistence changes when comparing the last two attribute changes.

In the Editor, the decision was made to consider initial edits as dirtying operations that create an undo level and enable saving functionality. For some reason, the server sends an empty initial edit with empty strings for the title, excerpt, and content of new posts. This results in another unnecessary undo level. This PR fixes this by ignoring initial edits that would not change any values of the post.

## How has this been tested?

It was verified that typing two words with a pause in between does not create an undo level and that new posts don't load with an undo level.

## Types of Changes

*Bug Fix:* Stop inserting unnecessary undo levels between words an on empty posts.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by ellatrix** [comment]
@epiqueras I added some e2e tests that should pass, and do pass before #16932. They are still failing with these changes though. I think one undo level has been removed, but for some reason there's still one being created with the first letter.

**Comment by epiqueras** [comment]
That's because the Block Editor is marking it as persistent (calling `onInput`). That was somehow ignored before. Could we make the Block Editor not do that?

**Comment by ellatrix** [comment]
Sure. Where is it doing that?

**Comment by epiqueras** [comment]
>with the first letter.

When you type that first character.

**Comment by ellatrix** [comment]
@epiqueras I meant in the code :)

**Comment by epiqueras** [comment]
No idea, I thought you might know, because of your `RichText`/writing flow expertise.

**Comment by ellatrix** [comment]
@epiqueras RichText doesn't mark the first character as persistent. It continuously calls `onChange` and `createUndoLevel` after 1s of no typing. The latter should eventually move somewhere else, so it only calls `onChange`. The only other case it creates an undo level is when it automatically changes the content based on an input rule, which doesn't seem to be the case here.

**Comment by epiqueras** [comment]
So, the issue is probably coming from somewhere else in the block editor.

In your tests, you are also seeing the first character passed through `onInput` right?

**Comment by epiqueras** [comment]
@ellatrix 

Well, it follows this logic:

https://github.com/WordPress/gutenberg/blob/dd7782a6dcdf05dbfa97bc8b3c78855940eb02ec/packages/block-editor/src/store/reducer.js#L341-L350

The first character's edit is not updating the same attribute as the last edit right?

How should we go about this?

I think the logic behind `MARK_LAST_CHANGE_AS_PERSISTENT` is kind of flawed. It should be a part of the next edit, not a standalone action. So when you type and stop for a while, the next edit will mark the last edit as persistent. With the current flow, the last edit gets marked as persistent, and the next character that comes in is also marked as persistent.

Also, do you have an example of a release where this worked correctly?

**Comment by ellatrix** [comment]
`git checkout 256a6745354c80a63de53c62dc0ec107eed00c30` (the commit before merging #16932). Everything worked fine there.

**Comment by ellatrix** [comment]
Should we not make sure this all works before the last Gutenberg release of WP 5.3?

**Comment by epiqueras** [comment]
>Should we not make sure this all works before the last Gutenberg release of WP 5.3?

@ellatrix I think we can merge this now with the latest changes?

**Comment by ellatrix** [comment]
Rebased the branch as I was getting some PHP errors with the current checkout of WP core.

**Comment by ellatrix** [comment]
Looks way better now.

Still seeing some inconsistencies.

1. Major: Create a new post. Create a new paragraph and type something in it. Save the draft. Undo the last change. The typing isn't undone. Instead, it seem like the saving is "undone"? The "Save Draft" button appears again, even though the content didn't change.
2. Minor: When I open the demo content, the undo button is active, even though I didn't touch the content. This wasn't the case before.

**Review by ellatrix — COMMENTED** [review]
Functionally it looks good. Would be great if the last e2e test case that I added could be fixed. Codewise I don't have any objections, but would be great if someone familiar with the `core-data` package could provide a second review.

File: packages/core-data/src/reducer.js
Comment (ellatrix) [inline packages/core-data/src/reducer.js]: What is this checking for? Would be great to inline comment.

File: packages/editor/src/store/actions.js
Comment (ellatrix) [inline packages/editor/src/store/actions.js]: The `raw` key seems very API specific. Given all the API abstractions, is this the right place to check for the `raw` key?

**Comment by ellatrix** [comment]
> I think the logic behind `MARK_LAST_CHANGE_AS_PERSISTENT` is kind of flawed. It should be a part of the next edit, not a standalone action. So when you type and stop for a while, the next edit will mark the last edit as persistent. With the current flow, the last edit gets marked as persistent, and the next character that comes in is also marked as persistent.

So this is all fine in the end? Anything that should be changed?

**Comment by ellatrix** [comment]
Since this seems more or less ready, and fixing important issues, I'd suggest we include it in Gutenberg 6.5.

**Comment by ellatrix** [comment]
420ead2 adds a test case for the third item listed in this comment: https://github.com/WordPress/gutenberg/pull/16932#issuecomment-526149034. This PR fixes the issue.

**Comment by youknowriad** [comment]
For the demo content being dirty. That was a decision we made on purpose after some discussions with @jasmussen and @epiqueras

**Comment by ellatrix** [comment]
Ok sure, I agree, that's just a minor thing.

**Comment by ellatrix** [comment]
I'm fine with reverting the test case introduced in d1817b6 for now if it would take too long to fix. It's good to have the other fixes merged.

**Comment by ellatrix** [comment]
With that test case I meant the first item in https://github.com/WordPress/gutenberg/pull/17259#issuecomment-531659784, not the demo content one. Would still be great to fix for 5.3 separately.

File: packages/core-data/src/reducer.js
Comment (epiqueras) [reply packages/core-data/src/reducer.js]: That the last two edits are not the same.

File: packages/editor/src/store/actions.js
Comment (epiqueras) [reply packages/editor/src/store/actions.js]: Yeah, it's not ideal. I think we should deprecate initial edits if possible.

File: packages/core-data/src/reducer.js
Comment (ellatrix) [reply packages/core-data/src/reducer.js]: I'm more wondering what the filter is for. :)

Comment (epiqueras) [reply packages/core-data/src/reducer.js]: When an edit is a function it's an optimization to avoid running some expensive operation. We can't rely on the function references being the same so we opt out of comparing them here.

**Review by ellatrix — APPROVED** [review]
In my opinion, feel free to merge. :)

File: packages/core-data/src/reducer.js
Comment (mcsf) [reply packages/core-data/src/reducer.js]: That could use an inline comment.

Comment (epiqueras) [reply packages/core-data/src/reducer.js]: I'm working on it with another fix.

Comment (epiqueras) [reply packages/core-data/src/reducer.js]: #17452

**Comment by epiqueras** [comment]
@ellatrix This fixes the other issue: #17452.

**Pull request merged**
